demo/lexer_demo.cpp

```diff
@@ -1,14 +1,111 @@
 #include <iostream>
+#include "fmt/core.h"
+#include "fmt/color.h"
+#include "strings.hpp"
 #include "lexer.hpp"
 #include "token.hpp"
+#include "strings.hpp"
 
-int main()
-{
-    std::cout << "Type\t\tPosition\t\tSrc\n";
-    std::vector<Token> tokens = lexer::read("./examples/ex4.rdd");
-    for (Token val : tokens)
-    {
-        std::cout << int(val.type) << "\t\t" << val.srcPos << "\t\t" << val.image << "\n";
+std::vector<std::string> g_TokTypeStr{
+    "Eof",
+    "Error",
+    "OneLineComment",
+    "VarDecl",
+    "TypeDecl",
+    "RoutineDecl",
+    "Is",
+    "IntegerType",
+    "RealType",
+    "BooleanType",
+    "RecordType",
+    "ArrayType",
+    "True",
+    "False",
+    "WhileLoop",
+    "ForLoop",
+    "LoopBegin",
+    "End",
+    "ReverseRange",
+    "InRange",
+    "If",
+    "Then",
+    "Else",
+    "AndLogic",
+    "OrLogic",
+    "XorLogic",
+    "LessComp",
+    "LeqComp",
+    "GreaterComp",
+    "GeqComp",
+    "EqComp",
+    "NeqComp",
+    "AssignmentOp",
+    "MultOp",
+    "DivOp",
+    "ModOp",
+    "AddOp",
+    "SubOp",
+    "IntegerLiteral",
+    "RealLiteral",
+    "Identifier",
+    "Dot",
+    "TwoDots",
+    "Comma",
+    "ParenOpen",
+    "ParenClose",
+    "BracketOpen",
+    "BracketClose",
+    "Colon",
+    "Semicolon",
+    "NewLine",
+};
+
+std::string to_string(lexer::TokenType type) {
+    return g_TokTypeStr[static_cast<int>(type)];
```

Comment:
If I understand correctly, this is to map between each token type and its string representation? If so, is an array really the best mapping method?

Reply:
Unfortunately, yes. In C++ there's no simple way to do this. The array indexing is the fastest method, here's only one level of indirection.

Reply:
What's wrong with an [`unordered_map`](https://en.cppreference.com/w/cpp/container/unordered_map)? It feels more robust against changes to the order of the lines in the source code

Reply:
There're two points:
1. It's slower.
2. It doesn't matter. This code is not the final product, it's just an "integration test". We won't change the enum as frequently (if ever will) as we did anyway.